Support for LLLVM 15 16

MartinNowack opened the pull request:
This PR adds support for LLVM 15 and 16.

There have been a multitude of required changes that are split into separate commits to simplify the whole process.

In a nutshell, we transitioned to support opaque pointers starting with LLVM 15 explicitly.
All the other commits handle additional moving parts in KLEE that had to be fixed.

<!--
Thank you for contributing to KLEE. We are looking forward to reviewing your PR. However, given the small number of active reviewers and our limited time, it might take a while to do so. We aim to get back to each PR within one month, and often do so within one week.

To review your PR, please add a summary of the proposed changes and ensure all items are fulfilled in the checklist above, by placing an "x" inside each applicable pair of brackets. More details about each item can be found in the [Developer's Guide](https://klee.github.io/docs/developers-guide/#pull-requests).
-->

## Summary: 


## Checklist:
- [x] The PR addresses a single issue.  If it can be divided into multiple independent PRs, please do so.
- [x] The PR is divided into a logical sequence of commits OR a single commit is sufficient.
- [x] There are no unnecessary commits (e.g. commits fixing issues in a previous commit in the same PR).
- [x] Each commit has a meaningful message documenting what it does.
- [x] All messages added to the codebase, all comments, as well as commit messages are spellchecked.
- [x] The code is commented OR not applicable/necessary.
- [x] The patch is formatted via clang-format OR not applicable (if explicitly overridden leave unchecked and explain).
- [x] There are test cases for the code you added or modified OR no such test cases are required.

ccadar commented:
@MartinNowack great to see this!  I hope we'll see it ready soon!

However, we should first merge #1349 , #1380 and #1663.

jirislaby commented:
I see:
```
[   35s] Failing Tests (3):
[   35s]     KLEE :: Feature/VarArgByVal.c
[   35s]     KLEE :: Runtime/POSIX/DirConsistency.c
[   35s]     KLEE :: Runtime/POSIX/SymFileConsistency.c
```
[Full log](https://github.com/klee/klee/files/13670398/log.txt)

ccadar commented:
@MartinNowack I think this one is next!  Do you want to rebase it to double-check that the FreeBSD checks pass, after the latest update?  If you do, I spotted two tiny issues: "pointerws" is misspelled in commit message of 561fff8e4ec4875d3b3b01f2ad78251b4238df18 and LLVM 17 seems to have been introduced prematurely in https://github.com/klee/klee/pull/1664/commits/5b13c5171adff19aeaebfa96ade25abd08f5a6c7.

danielschemmel commented on scripts/build/p-libcxx.inc:
```diff
@@ -53,7 +59,11 @@ build_libcxx() {
     cmake "${cmake_flags[@]}" "${LIBCXX_SRC}/llvm"
-    make cxx -j$(nproc) || make cxx
+   if [[ "${LLVM_VERSION_SHORT}" -ge "14" ]]; then
+     make runtimes - j"$(nproc)" || make cxx || return 1
+   else
+     make cxx - j"$(nproc)" || make cxx || return 1
```
This looks like the original intent was to run a parallel make followed by a serial make of the same target. This change introduces two problems:
1. There is a blank between the `-` and the `j` that should not be there
2. it should probably also run `make runtimes` in the fallback position of the then-case

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -1520,18 +1527,16 @@ MemoryObject *Executor::serializeLandingpad(ExecutionState &state,
+            clause_value = dyn_cast<GlobalValue>(bitcast->getOperand(0));
+          }
+
+          if (auto *gv = dyn_cast<llvm::GlobalVariable>(v)) {
```
```suggestion
          if (auto const *gv = dyn_cast<llvm::GlobalVariable>(v)) {
```

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -3405,15 +3437,66 @@ void Executor::computeOffsets(KGEPInstruction *kgepi, TypeIt ib, TypeIt ie) {
+#else
+    llvm::Value *agg = ivi->getAggregateOperand();
+    llvm::Type *current_type = agg->getType();
+    uint64_t offset = 0;
```
```suggestion
    std::uint64_t offset = 0;
```

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -3405,15 +3437,66 @@ void Executor::computeOffsets(KGEPInstruction *kgepi, TypeIt ib, TypeIt ie) {
+    for (auto index : ivi->indices()) {
+      if (StructType *st = dyn_cast<llvm::StructType>(current_type)) {
+        const StructLayout *sl = kmodule->targetData->getStructLayout(st);
+        uint64_t addend = sl->getElementOffset(index);
```
```suggestion
        std::uint64_t addend = sl->getElementOffset(index);
```

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -3405,15 +3437,66 @@ void Executor::computeOffsets(KGEPInstruction *kgepi, TypeIt ib, TypeIt ie) {
+        offset = offset + addend;
+      } else if (current_type->isArrayTy() || current_type->isVectorTy() ||
+                 current_type->isPointerTy()) {
+        uint64_t elementSize = kmodule->targetData->getTypeStoreSize(
```
```suggestion
        std::uint64_t elementSize = kmodule->targetData->getTypeStoreSize(
```

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -3405,15 +3437,66 @@ void Executor::computeOffsets(KGEPInstruction *kgepi, TypeIt ib, TypeIt ie) {
+
+    llvm::Value *agg = evi->getAggregateOperand();
+    llvm::Type *current_type = agg->getType();
+    uint64_t offset = 0;
```
```suggestion
    std::uint64_t offset = 0;
```

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -3405,15 +3437,66 @@ void Executor::computeOffsets(KGEPInstruction *kgepi, TypeIt ib, TypeIt ie) {
+    for (auto index : evi->indices()) {
+      if (StructType *st = dyn_cast<llvm::StructType>(current_type)) {
+        const StructLayout *sl = kmodule->targetData->getStructLayout(st);
+        uint64_t addend = sl->getElementOffset(index);
```
```suggestion
        std::uint64_t addend = sl->getElementOffset(index);
```

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -3405,15 +3437,66 @@ void Executor::computeOffsets(KGEPInstruction *kgepi, TypeIt ib, TypeIt ie) {
+        offset = offset + addend;
+      } else if (current_type->isArrayTy() || current_type->isVectorTy() ||
+                 current_type->isPointerTy()) {
+        uint64_t elementSize = kmodule->targetData->getTypeStoreSize(
```
```suggestion
        std::uint64_t elementSize = kmodule->targetData->getTypeStoreSize(
```

danielschemmel commented on lib/Module/Instrument.cpp:
```diff
@@ -0,0 +1,19 @@
+
+using namespace klee;
+
+void klee::checkModule(bool DontVerfify, llvm::Module *module) { assert(0); }
```
```suggestion
void klee::checkModule(bool DontVerify, llvm::Module *module) { assert(0); }
```

danielschemmel commented on lib/Module/Instrument.cpp:
```diff
@@ -0,0 +1,19 @@
+
+#include "ModuleHelper.h"
+
+using namespace klee;
```
```suggestion
```

(All names are fully qualified)

danielschemmel commented on lib/Module/Optimize.cpp:
```suggestion
using namespace klee;

void klee::optimiseAndPrepare(bool OptimiseKLEECall, bool Optimize,
```

danielschemmel commented on test/Feature/BFSSearcherAndDFSSearcherInterleaved.c:
```diff
@@ -13,7 +13,7 @@
 
 #include "klee/klee.h"
-
+#include <stdio.h>
```
```suggestion
#include <stdio.h>

```

danielschemmel commented on test/Feature/MultipleReadResolution.c:
```diff
@@ -7,8 +7,9 @@
+#include "klee/klee.h"
 #include <stdio.h>
-
+#include <stdlib.h>
 unsigned klee_urange(unsigned start, unsigned end) {
```
```suggestion
#include <stdlib.h>

unsigned klee_urange(unsigned start, unsigned end) {
```

danielschemmel commented on test/Feature/MultipleWriteResolution.c:
```diff
@@ -6,9 +6,9 @@
+#include "klee/klee.h"
 #include <stdio.h>
-
+#include <stdlib.h>
 unsigned klee_urange(unsigned start, unsigned end) {
```
```suggestion
#include <stdlib.h>

unsigned klee_urange(unsigned start, unsigned end) {
```

danielschemmel commented on test/Feature/consecutive_divide_by_zero.c:
```diff
@@ -8,6 +8,8 @@
 *  DISTINCT ERRORS!
 */
+
+#include "klee/klee.h"
 int main() {
```
```suggestion
#include "klee/klee.h"

int main() {
```

danielschemmel commented on test/Runtime/POSIX/FreeArgv.c:
```diff
@@ -4,7 +4,8 @@
 // RUN: test -f %t.klee-out/test000003.free.err
-
+#include "klee/klee.h"
+#include <stdlib.h>
 int main(int argc, char **argv) {
```
```suggestion
#include <stdlib.h>

int main(int argc, char **argv) {
```

danielschemmel commented on test/Runtime/POSIX/Getenv.c:
```diff
@@ -1,9 +1,10 @@
-
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
 int main(int argc, char **argv) {
```
```suggestion
#include <string.h>

int main(int argc, char **argv) {
```

danielschemmel commented on test/regression/2007-08-16-valid-write-to-freed-object.c:
```diff
@@ -2,6 +2,8 @@
 // RUN: %klee --output-dir=%t.klee-out %t1.bc
 
+#include "klee/klee.h"
+#include <stdio.h>
 unsigned sym() {
```
```suggestion
#include <stdio.h>

unsigned sym() {
```

danielschemmel commented on test/regression/2007-10-11-free-of-alloca.c:
```diff
@@ -2,7 +2,7 @@
 // RUN: %klee --output-dir=%t.klee-out %t1.bc 2>&1 | FileCheck %s
 // RUN: test -f %t.klee-out/test000001.free.err
-
+#include <stdlib.h>
 int main() {
```
```suggestion
#include <stdlib.h>

int main() {
```

danielschemmel commented on test/regression/2007-10-12-failed-make-symbolic-after-copy.c:
```diff
@@ -2,7 +2,7 @@
 // RUN: %klee --output-dir=%t.klee-out %t1.bc
 // RUN: test -f %t.klee-out/test000001.ktest
-
+#include "klee/klee.h"
 int main() {
```
```suggestion
#include "klee/klee.h"

int main() {
```

danielschemmel commented on test/regression/2008-03-04-free-of-global.c:
```diff
@@ -3,10 +3,11 @@
 // RUN: %klee --output-dir=%t.klee-out %t1.bc 2>&1 | FileCheck %s
 // RUN: test -f %t.klee-out/test000001.free.err
 
+#include <stdlib.h>
 int buf[4];
```
```suggestion
#include <stdlib.h>

int buf[4];
```

danielschemmel commented on test/regression/2008-04-10-bad-alloca-free.c:
```diff
@@ -2,6 +2,7 @@
 // RUN: rm -rf %t.klee-out
 // RUN: %klee --output-dir=%t.klee-out --exit-on-error %t1.bc
 
+#include "klee/klee.h"
 void f(int *addr) {
```
```suggestion
#include "klee/klee.h"

void f(int *addr) {
```

danielschemmel commented on test/regression/2014-07-04-unflushed-error-report.c:
```diff
@@ -6,6 +6,7 @@
 /* This test checks that the error file isn't empty and contains the
  * right content.
  */
+#include "klee/klee.h"
 int main() {
```
```suggestion
#include "klee/klee.h"

int main() {
```

danielschemmel commented on test/regression/2015-08-30-empty-constraints.c:
```diff
@@ -10,6 +10,7 @@
  * are generated.
  * Make sure we are able to generate an input.
  */
+#include "klee/klee.h"
 int main() {
```
```suggestion
#include "klee/klee.h"

int main() {
```

danielschemmel commented on test/regression/2015-08-30-sdiv-1.c:
```diff
@@ -7,6 +7,7 @@
  * For signed division, div by 1 or -1 cannot be optimized like that.
  */
+#include "klee/klee.h"
 #include <stdint.h>
 int main() {
```
```suggestion
#include <stdint.h>

int main() {
```

danielschemmel commented on test/regression/2017-02-21-pathOS-id.c:
```diff
@@ -5,6 +5,9 @@
 // RUN: cat %t.klee-out/test000004.path | wc -l | grep -q 1
+
+#include "klee/klee.h"
+#include "stdlib.h"
 int main(){
```
```suggestion
#include "stdlib.h"

int main(){
```

danielschemmel commented on test/regression/2017-11-01-test-with-empty-varname.c:
```diff
@@ -2,7 +2,7 @@
 // RUN: %klee --output-dir=%t.klee-out %t1.bc
 // RUN: FileCheck %s --input-file=%t.klee-out/warnings.txt
-
+#include "klee/klee.h"
 int main() {
```
```suggestion
#include "klee/klee.h"

int main() {
```

danielschemmel reviewed (commented):
Thank you very much for all your hard work on this, @MartinNowack!

I have a few comments, most of which are just stylistic (`uint64_t` is actually called `std::uint64_t` and for most of the codebase we have an empty line between the includes and the rest of the code), but also one comment about the cmake changes that looks somewhat important.

I hope you can either just one-click apply or ignore the many small suggestions ;).

danielschemmel commented on lib/Core/Executor.cpp:
```diff
@@ -3365,10 +3382,17 @@ template <typename TypeIt>
+#else
+  assert(it.isSequential() && "Called with non-sequential type");
+  // Get the size of a single element
+  uint64_t elementSize =
```
```suggestion
  std::uint64_t elementSize =
```

danielschemmel commented on lib/Module/OptimizeLegacy.cpp:
```diff
@@ -0,0 +1,259 @@
+
+static cl::opt<bool>
+    StripDebug("strip-debug",
+               cl::desc("Strip debugger symbol info from executable"),
```
```suggestion
               cl::desc("Strip debugger symbol info from executable (default=false)"),
```

danielschemmel commented on lib/Module/OptimizeLegacy.cpp:
```diff
@@ -0,0 +1,259 @@
+                                cl::init(false), cl::cat(klee::ModuleCat));
+
+static cl::opt<bool>
+    Strip("strip-all", cl::desc("Strip all symbol information from executable"),
```
```suggestion
    Strip("strip-all", cl::desc("Strip all symbol information from executable (default=false)"),
```

MartinNowack commented:
Thanks @ccadar and @danielschemmel for the review. I addressed most of the comments, I kept the `lit`'s one LLVM 17. We will need it anyway soon 😃  So, this one is ready!

ccadar commented:
Thanks again, @MartinNowack , for the hard work of adding support for LLVM 15 & 16!

ccadar commented:
Thanks, @jirislaby, I somehow thought that was resolved. Can you please open a new issue with more details? (what platform and configuration you are using, how the tests fail)?
